test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java

```diff
@@ -2145,6 +2153,85 @@ public static <T> T safeAwait(SubscribableListener<T> listener) {
         }
     }
 
+    /**
+     * Waits for async function to complete with {@value #AWAIT_LISTENER_TIMEOUT} sec timeout.
+     * Can be used when you expect a result of async operation.
+     * For example:
+     * <pre>{@code
+     *         final var resp = safeAwait(
+     *             (ActionListener<AcknowledgedResponse> listener) -> repositoriesService.registerRepository(req, listener)
+     *         );
+     *         assertTrue(resp.isAcknowledged());
+     * }</pre>
+     *
+     * @return result of listener onResponse
+     * @throws AssertionError if listener completed with exception, or thread interrupted, or timed out
+     */
+    public static <T> T safeAwait(Consumer<ActionListener<T>> asyncFunc) {
+        return safeAwait(SubscribableListener.newForked(asyncFunc::accept));
+    }
+
+    /**
+     * Waits for async function to complete with {@value #AWAIT_LISTENER_TIMEOUT} sec timeout.
+     * Can be used when you expect a failure of async operation.
+     * For example:
+     * <pre>{@code
+     *         final var errResp = safeAwaitFailure(
+     *             (ActionListener<AcknowledgedResponse> listener) -> repositoriesService.registerRepository(req, listener)
+     *         );
+     *         assertThat(errResp, isA(RepositoryVerificationException.class));
+     * }</pre>
+     *
+     * @return result of listener onFailure
+     * @throws AssertionError if listener completed successfully, or thread interrupted, or timed out
+     */
+    public static <T> Exception safeAwaitFailure(Consumer<ActionListener<T>> asyncFunc) {
+        return safeAwaitFailure(SubscribableListener.newForked(asyncFunc::accept));
+    }
+
+    /**
+     * Waits for listener to complete with {@value #AWAIT_LISTENER_TIMEOUT} sec timeout.
+     * Can be used when you expect a failure of async operation.
+     *
+     * @return result of listener onFailure
+     * @throws AssertionError if listener completed successfully, or thread interrupted, or timed out
+     */
+    public static <T> Exception safeAwaitFailure(SubscribableListener<T> listener) {
```

Comment:
This approach introduces quite a lot of machinery for catching a failure like this, I suspect we can do the same thing by composing (mostly) existing utilities, plus the one introduced in https://github.com/elastic/elasticsearch/pull/108547.

Reply:
I couldnt figure out how to avoid try/catch without changing safeAwait. In safeAwait we throw all exceptions(listener and timeout) as AssertException, then I need another try/catch in safeAwaitFaliure to unwrap AssertException. It saves few lines of code, but not obviously better.

Using `assertNoFailureListener` in awaitFailure still requires try/catch from above. Do you have a suggestion how to compose existing methods?

My thinking about composition is changing `safeAwait` to return a tuple of (Response, Failure), and then have `safeAwaitResponse`/`safeAwaitFailure` related methods. Names and meaning are composed accordingly.

```java
record ListenerResult<T>(T response, Exception failure){}

public static <T> ListenerResult<T> safeAwait(SubscribableListener<T> listener) {
    try {
        final var future = new PlainActionFuture<T>();
        listener.addListener(future);
        final var response = future.get(timeout, timeUnit);
        return new ListenerResult(response, null);
    } catch (ExecutionException e) {
        return new ListenerResult(null, e.getCause());
    } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new AssertionError("safeAwait: interrupted waiting for SubscribableListener", e);
    } catch (TimeoutException e) {
        throw new AssertionError("safeAwait: listener was not completed within the timeout", e);
    }
}

public static <T> T safeAwaitResponse(SubscribableListener<T> listener) {
    final var result = safeAwait(listener);
    assertNull(result.failure);
    assertNotNull(result.response);
    return result.response;
}

public static <T> T safeAwaitFailure(SubscribableListener<T> listener) {
    final var result = safeAwait(listener);
    assertNull(result.response);
    assertNotNull(result.failure);
    return result.failure;
}

```

Reply:
Or even this, more type safety, like rust enums. "Switch" statement on sealed interface enforce exhaustive check on branches at compile time.

```java
    sealed interface ListenerResult<T> permits ListenerResponse, ListenerFailure {};

    record ListenerResponse<T>(T response) implements ListenerResult<T> {}

    record ListenerFailure(Throwable failure) implements ListenerResult {}

    public static <T> ListenerResult<T> safeAwait(SubscribableListener<T> listener) {
        try {
            final var future = new PlainActionFuture<T>();
            listener.addListener(future);
            final var response = future.get(timeout, timeUnit);
            return new ListenerResponse<>(response);
        } catch (ExecutionException e) {
            return new ListenerFailure(e.getCause());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new AssertionError("safeAwait: interrupted waiting for SubscribableListener", e);
        } catch (TimeoutException e) {
            throw new AssertionError("safeAwait: listener was not completed within the timeout", e);
        }
    }

    public static <T> T safeAwaitResponse(SubscribableListener<T> listener) {
        final var result = safeAwait1(listener);
        switch (result) {
            case ListenerFailure listenerFailure -> {
                throw new AssertionError(listenerFailure.failure);
            }
            case ListenerResponse<T> listenerResponse -> {
                return listenerResponse.response;
            }
        }
    }

    public static <T> Throwable safeAwaitFailure(SubscribableListener<T> listener) {
        final var result = safeAwait1(listener);
        switch (result) {
            case ListenerFailure listenerFailure -> {
                return listenerFailure.failure;
            }
            case ListenerResponse<T> listenerResponse -> {
                throw new AssertionError("must fail");
            }
        }
    }

```

Reply:
This all still seems rather overcomplicated, I think we can write the test something like this:

```
    public void testRegisterThrowsVerificationFailureBeforeClusterUpdate() {
        final var req = new PutRepositoryRequest("verify-fail-repo");
        req.type(VerifyFailRepository.TYPE);
        final var exception = safeAwait(
            SubscribableListener.<Exception>newForked(
                l -> repositoriesService.registerRepository(req, ActionTestUtils.assertNoSuccessListener(l::onResponse))
            )
        );
        assertThat(exception, isA(RepositoryVerificationException.class));
        assertEquals(0, repositoriesService.getRepositories().size()); // assert cluster is not updated
    }
```

Maybe there's a utility to be extracted there but I don't think we need all this extra code unless I'm missing something?

Reply:
That works, probably personal preference. I would prefer shorter meaningful methods like awaitFailure/awaitResponse. It's easier to show intention of the code.

```java
        final var exception = safeAwait(
            SubscribableListener.<Exception>newForked(
                l -> repositoriesService.registerRepository(req, ActionTestUtils.assertNoSuccessListener(l::onResponse))
            )
        );
```

I would prefer less callback-y. My body is still adjusting to callbacks, there are changes on molecular level :D

```java
final var listener = new SubscribableListener<AcknowledgedResponse>();
repositoriesService.registerRepository(req, listener);
final var failure = safeAwaitFailure(listener);
assertThat(exception, isA(RepositoryVerificationException.class));
```